From c89a2dc1245a2ddd17fadb643ce4c7654bcbb054 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Sat, 26 Oct 2013 07:31:05 +0800
Subject: [PATCH] Add XOR_PREFER_TEMPLATE to xor[v2].

Use XOR_PREFER_TEMPLATE can avoid xor template checking at bootup.
---
 arch/x86/include/asm/xor.h    | 2 +-
 arch/x86/include/asm/xor_32.h | 3 +++
 arch/x86/include/asm/xor_64.h | 2 ++
 3 files changed, 6 insertions(+), 1 deletion(-)

diff --git a/arch/x86/include/asm/xor.h b/arch/x86/include/asm/xor.h
index d882975..dd2b820 100644
--- a/arch/x86/include/asm/xor.h
+++ b/arch/x86/include/asm/xor.h
@@ -494,6 +494,6 @@ static struct xor_block_template xor_block_sse_pf64 = {
 #endif
 
 #define XOR_SELECT_TEMPLATE(FASTEST) \
-	AVX_SELECT(FASTEST)
+	AVX_SELECT((NULL==(FASTEST))? XOR_PREFER_TEMPLATE:FASTEST)
 
 #endif /* _ASM_X86_XOR_H */
diff --git a/arch/x86/include/asm/xor_32.h b/arch/x86/include/asm/xor_32.h
index ce05722..0c2910f 100644
--- a/arch/x86/include/asm/xor_32.h
+++ b/arch/x86/include/asm/xor_32.h
@@ -564,4 +564,7 @@ do {							\
 	}						\
 } while (0)
 
+/* We doesn't prefer any block so to trigger the try */
+#define XOR_PREFER_TEMPLATE (NULL)
+
 #endif /* _ASM_X86_XOR_32_H */
diff --git a/arch/x86/include/asm/xor_64.h b/arch/x86/include/asm/xor_64.h
index 546f1e3..0e0f7ef 100644
--- a/arch/x86/include/asm/xor_64.h
+++ b/arch/x86/include/asm/xor_64.h
@@ -24,4 +24,6 @@ do {						\
 	xor_speed(&xor_block_sse);		\
 } while (0)
 
+/* We prefer SSE xor pf64, AVX is optional */
+#define XOR_PREFER_TEMPLATE (&xor_block_sse_pf64)
 #endif /* _ASM_X86_XOR_64_H */
